File: frontend/src/services/recuPerceptionService.ts
import { get, post, put, del } from "./api";

export type TypePerception = 'peage_urbain' | 'pont_bascule' | 'etalage' | 'chargement' | 'dechargement' | 'autre';

export const typePerceptionLabels: Record<TypePerception, string> = {
  peage_urbain: 'Péage urbain',
  pont_bascule: 'Pont bascule',
  etalage: 'Étalage',
  chargement: 'Chargement',
  dechargement: 'Déchargement',
  autre: 'Autre',
};

export const typePerceptionPrefix: Record<TypePerception, string> = {
  peage_urbain: 'PEA',
  pont_bascule: 'PON',
  etalage: 'ETA',
  chargement: 'CHA',
  dechargement: 'DEC',
  autre: 'AUT',
};

export interface RecuPerception {
  id: number;
  taxe_id: number;
  personne_id: number | null;
  percepteur_id: number | null;
  type_perception: string;
  numero: string;
  date_emission: string;
  heure_emission: string | null;
  categorie_vehicule: string | null;
  plaque_immatriculation: string | null;
  montant: number;
  trajet: 'aller' | 'retour' | 'aller_retour' | null;
  chauffeur_nom: string | null;
  conducteur_nom: string | null;
  Numero_Piece: string | null;
  designation: string | null;
  poids: number | null;
  observations: string | null;
  taxe?: { id: number; nom: string; code: string };
  personne?: { id: number; nom: string; prenom: string; nom_complet: string };
  percepteur?: { id: number; nom: string; prenom: string };
  created_at: string;
}

export interface RecuPerceptionCreate {
  taxe_id: number;
  personne_id?: number;
  type_perception: string;
  date_emission: string;
  heure_emission?: string;
  categorie_vehicule?: string;
  plaque_immatriculation?: string;
  montant: number;
  trajet?: 'aller' | 'retour' | 'aller_retour';
  chauffeur_nom?: string;
  conducteur_nom?: string;
  Numero_Piece?: string;
  designation?: string;
  poids?: number;
  observations?: string;
}

export interface PaginatedRecus {
  data: RecuPerception[];
  current_page: number;
  per_page: number;
  total: number;
  last_page: number;
}

export const recuPerceptionService = {
  list(params?: { page?: number; per_page?: number; type_perception?: string; date_debut?: string; date_fin?: string; taxe_id?: number; search?: string }): Promise<PaginatedRecus> {
    return get<PaginatedRecus>('/recus-perception', params as Record<string, string | number | boolean | undefined>);
  },

  get(id: number): Promise<RecuPerception> {
    return get<RecuPerception>(`/recus-perception/${id}`);
  },

  create(data: RecuPerceptionCreate): Promise<RecuPerception> {
    return post<RecuPerception>('/recus-perception', data);
  },

  update(id: number, data: Partial<RecuPerceptionCreate>): Promise<RecuPerception> {
    return put<RecuPerception>(`/recus-perception/${id}`, data);
  },

  delete(id: number): Promise<void> {
    return del<void>(`/recus-perception/${id}`);
  },

  prochainNumero(typePerception?: string): Promise<{ numero: string }> {
    return get<{ numero: string }>('/recus-perception/prochain-numero', { type_perception: typePerception } as Record<string, string>);
  },

  stats(params?: { type_perception?: string; date_debut?: string; date_fin?: string }): Promise<{ total_recus: number; montant_total: number; par_type: Array<{ type_perception: string; total: number; montant: number }> }> {
    return get('/recus-perception/stats', params as Record<string, string | number | boolean | undefined>);
  },

  types(): Promise<Record<string, string>> {
    return get<Record<string, string>>('/recus-perception/types');
  },

  qrcode(id: number): Promise<{ numero: string; url: string }> {
    return get<{ numero: string; url: string }>(`/recus-perception/${id}/qrcode`);
  },
};

File: gs_operateur_mobile/lib/services/recu_perception_service.dart
import 'package:dio/dio.dart';
import '../models/recu_perception.dart';
import 'api_service.dart';

class RecuPerceptionService {
  final ApiService _api = ApiService();

  Future<List<RecuPerception>> list({String? search, String? typePerception, int page = 1}) async {
    final response = await _api.get('/recus-perception', queryParameters: {
      if (search != null && search.isNotEmpty) 'search': search,
      if (typePerception != null) 'type_perception': typePerception,
      'page': page,
    });

    if (response['success'] == true) {
      final data = response['data'];
      final items = data is Map ? (data['data'] ?? []) : data;
      return (items as List).map<RecuPerception>((r) => RecuPerception.fromJson(r)).toList();
    }
    throw Exception(response['message'] ?? 'Erreur');
  }

  Future<RecuPerception> get(int id) async {
    final response = await _api.get('/recus-perception/$id');
    if (response['success'] == true) {
      return RecuPerception.fromJson(response['data']);
    }
    throw Exception(response['message'] ?? 'Erreur');
  }

  Future<RecuPerception> create(Map<String, dynamic> data) async {
    final response = await _api.post('/recus-perception', data: data);
    if (response['success'] == true) {
      return RecuPerception.fromJson(response['data']);
    }
    throw Exception(response['message'] ?? 'Erreur lors de la création');
  }

  Future<RecuPerception> update(int id, Map<String, dynamic> data) async {
    final response = await _api.put('/recus-perception/$id', data: data);
    if (response['success'] == true) {
      return RecuPerception.fromJson(response['data']);
    }
    throw Exception(response['message'] ?? 'Erreur lors de la modification');
  }

  Future<void> delete(int id) async {
    final response = await _api.delete('/recus-perception/$id');
    if (response['success'] != true) {
      throw Exception(response['message'] ?? 'Erreur lors de la suppression');
    }
  }

  Future<Map<String, dynamic>> stats() async {
    final response = await _api.get('/recus-perception/stats');
    if (response['success'] == true) {
      return response['data'];
    }
    throw Exception(response['message'] ?? 'Erreur');
  }

  Future<List<Map<String, dynamic>>> types() async {
    final response = await _api.get('/recus-perception/types');
    if (response['success'] == true) {
      return List<Map<String, dynamic>>.from(response['data']);
    }
    throw Exception(response['message'] ?? 'Erreur');
  }

  Future<String> prochainNumero(String typePerception) async {
    final response = await _api.get('/recus-perception/prochain-numero', queryParameters: {
      'type_perception': typePerception,
    });
    if (response['success'] == true) {
      return response['data']['numero'];
    }
    throw Exception(response['message'] ?? 'Erreur');
  }

  Future<Map<String, dynamic>> qrcode(int id) async {
    final response = await _api.get('/recus-perception/$id/qrcode');
    if (response['success'] == true) {
      return response['data'];
    }
    throw Exception(response['message'] ?? 'Erreur');
  }
}
